### PR TITLE
feat: add method for getting the version of the running quilc server

### LINKED ISSUE #205: Python: Add method for getting version info from quilc

### PR DESCRIPTION
**Pull request by MarquessV**
Closes #205

### CODE REVIEW

File: crates/lib/src/qpu/quilc/mod.rs [hunk 1]
```diff
@@ -50,6 +50,19 @@ pub(crate) fn compile_program(
     }
 }
 
+pub(crate) fn get_version_info(client: &Qcs) -> Result<String, Error> {
+    let config = client.get_config();
+    let endpoint = config.quilc_url();
+    let binding: HashMap<&str, &str> = HashMap::new();
+    let request = rpcq::RPCRequest::new("get_version_info", &binding);
```
Comment (MarquessV) [inline crates/lib/src/qpu/quilc/mod.rs:57]: What is the best way to pass in an empty set of params to `rpcq::RPCRequest::new`? I don't love initializing this empty `HashMap` just for that purpose, but the function needs something, and the empty types I tried crash quilc. cc'ing: @Shadow53 for their Rust expertise.

Comment (Shadow53) [reply crates/lib/src/qpu/quilc/mod.rs:57]: `HashMap::new` should be fine. The standard library generally tries to avoid allocating until necessary, which [is the case](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#method.new) with `HashMap::new`.

Could also try passing in `()` if you haven't already. I think that implements `Serialize`. Only other option I can think to try is passing `None`, but I'm pretty sure the compiler would complain about not knowing what type the `Option` should contain.

File: crates/python/tests/test_api.py [hunk 2]
```diff
@@ -1,11 +1,12 @@
 from unittest.mock import patch
 import pytest
 
+import re
+
 import qcs_sdk
 
 
 @pytest.mark.asyncio
-@pytest.mark.skip
```
Comment (Shadow53) [inline crates/python/tests/test_api.py:8]: This seems unrelated to getting quilc version. Was this change intended?

**Review by Shadow53 — CHANGES_REQUESTED** [review]

File: crates/lib/src/qpu/quilc/mod.rs (hunk 1 above)
Comment (MarquessV) [reply crates/lib/src/qpu/quilc/mod.rs:57]: I tried `()` first, which also unfortunately doesn't work. Rust doesn't mind, but `quilc` throws a pretty gnarly error, dies, and leaves the Rust library hanging.

```
Unhandled TYPE-ERROR in thread #<SB-THREAD:THREAD "RPC-server-thread-0" RUNNING
                                  {1005A834B3}>:
  The value
    NIL
  is not of type
    HASH-TABLE
```

File: crates/python/tests/test_api.py (hunk 2 above)
Comment (MarquessV) [reply crates/python/tests/test_api.py:8]: Yes, when I added a test for the new quilc method, I also updated `conftest.py` so that these tests will load the config from `tests/qcs_config`. That way the test would work without having a QCS config in `~/.qcs` and will work in CI. That change also allowed this test to run successfully on CI so I removed the `skip` decoration. 

The other tests with `skip` would require a valid QCS config so we can't run those on CI just yet.

File: crates/lib/src/api.rs [hunk 3]
```diff
@@ -34,6 +34,11 @@ pub fn compile(
         .map(|p| p.to_string(true))
 }
 
+/// Gets the quilc version
+pub fn get_quilc_version(client: &Qcs) -> Result<String, Box<dyn std::error::Error>> {
```
[the remark this answers is not in the document]
Comment (MarquessV) [reply crates/lib/src/api.rs:38]: The compiler doesn't like that:

image

File: crates/lib/src/qpu/quilc/mod.rs (hunk 1 above)
Comment (Shadow53) [reply crates/lib/src/qpu/quilc/mod.rs:57]: Sounds good. Good to know that it requires a hash map.

It's strictly unrelated to this change, but it would be a good idea to update the documentation and/or signature for making a new `RPCRequest` to only allow hash maps, e.g. using (pseudo code) `binding: HashMap<K=String, V=String>` rather than `binding: T = HashMap<String, String>`

File: crates/lib/src/api.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Shadow53) [reply crates/lib/src/api.rs:38]: If `Sync` is the only thing failing, that's fine to remove. I couldn't remember if it's something usually used here or not -- the more important ones are `'static` and `Send`.

[the remark this answers is not in the document]
Comment (MarquessV) [reply crates/lib/src/api.rs:38]: It'll take `Error + 'static` but not ` + Send` or ` + Sync`

File: crates/lib/src/qpu/quilc/mod.rs (hunk 1 above)
Comment (MarquessV) [reply crates/lib/src/qpu/quilc/mod.rs:57]: I like the idea of hardening the type that gets accepted, but I think only allowing HashMap would be too restrictive. For example, the compile function uses the `QuilcParams` struct. We only use a HashMap here as a quick way of creating an empty set of params.

File: crates/lib/src/api.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (MarquessV) [reply crates/lib/src/api.rs:38]: Michael and I discussed this offline - he found the issue within quil-rs:

```
1. Box has two From impls related to errors: From<E> for Box<dyn Error + 'a> and From<E> for Box<dyn Error + Send + Sync + 'a>. That is, there's no impl for Box<dyn Error + Send + 'a.
2. quilc::Error is not Sync because quil_rs::program::SyntaxError is not Sync.
3. SyntaxError is not Sync because quil_rs::parser::error::error::Error is not Sync
4. quil_rs::parser::error::error::Error is not Sync because it contains a field of type Option<Box<dyn Error + Send + 'static>>.
```

We have to add `Sync` to the parser error in quil-rs, then we can do it here. Adding a TODO since there is another method in this API with the same problem.

File: crates/lib/src/api.rs [hunk 4]
```diff
@@ -24,16 +24,23 @@ use crate::qpu::{
 };
 
 /// Uses quilc to convert a Quil program to native Quil
+/// TODO: Add `+ Send + Sync` to the error type once quil-rs supports it: <https://github.com/rigetti/quil-rs/issues/122>
```
Comment (Shadow53) [inline crates/lib/src/api.rs:27]: Should add a `TODO` for this repo as well.

File: crates/lib/src/api.rs [hunk 5]
```diff
@@ -24,16 +24,23 @@ use crate::qpu::{
...
     client: &Qcs,
-) -> Result<String, Box<dyn std::error::Error>> {
+) -> Result<String, Box<dyn std::error::Error + 'static>> {
     quilc::compile_program(quil, target, client)
         .map_err(Into::into)
         .map(|p| p.to_string(true))
 }
 
+/// Gets the quilc version
+/// TODO: Add `+ Send + Sync` to the error type once quil-rs supports it: <https://github.com/rigetti/quil-rs/issues/122>
```
Comment (Shadow53) [inline crates/lib/src/api.rs:39]: Same here

**Review by Shadow53 — CHANGES_REQUESTED** [review]
Add links to an issue in this repository, then this looks good.

**Review by Shadow53 — APPROVED** [review]
lgtm
